From jasonbstubbs@gmail.com  Tue May 10 15:37:55 2011
From: Jason Stubbs <jasonbstubbs@gmail.com>
To: Greg KH <gregkh@suse.de>
Subject: platform: samsung_laptop: add support for samsung nc210/nc110
Date: Wed, 20 Apr 2011 15:07:54 +1000
Cc: linux-kernel@vger.kernel.org
Message-Id: <201104201507.54862.jasonbstubbs@gmail.com>

From: Jason Stubbs <jasonbstubbs@gmail.com>

Note, I don't expect this patch to go in and would actually just like some
feedback on how to better handle the set_brightness problem. Note also that
this patch won't apply on top of my brightness min/max fixes patch and I'll
redo this one as necessary.

This patch adds support for the NC210/NC110 to the samsung-laptop driver.

All SABI commands except for set_brightness work as expected. The behaviour
of set_brightness is as follows:

- Setting a new brightness will only step one level toward the new brightness
  level. For example, setting a level of 5 when the current level is 2 will
  result in a brightness level of 3.
- A spurious KEY_BRIGHTNESS_UP or KEY_BRIGHTNESS_DOWN event is also generated
  along with the change in brightness.
- Neither of the above two issues occur when changing from/to brightness level 0.

The changes in set_brightness() work around the above while attempting to
minimize flicker due to jumping to and from brightness level 0.

Signed-off-by: Jason Stubbs <jasonbstubbs@gmail.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/platform/x86/samsung-laptop.c |   15 +++++++++++++++
 1 file changed, 15 insertions(+)

--- a/drivers/platform/x86/samsung-laptop.c
+++ b/drivers/platform/x86/samsung-laptop.c
@@ -380,6 +380,12 @@ static void set_brightness(u8 user_brigh
 {
 	u8 user_level = user_brightness - sabi_config->min_brightness;
 
+	if (user_level == read_brightness())
+		return;
+
+	if (sabi_config->min_brightness == 0 && user_level != 0)
+		sabi_set_command(sabi_config->commands.set_brightness, 0);
+
 	sabi_set_command(sabi_config->commands.set_brightness, user_level);
 }
 
@@ -675,6 +681,15 @@ static struct dmi_system_id __initdata s
 		},
 		.callback = dmi_check_cb,
 	},
+	{
+		.ident = "NC210/NC110",
+		.matches = {
+			DMI_MATCH(DMI_SYS_VENDOR, "SAMSUNG ELECTRONICS CO., LTD."),
+			DMI_MATCH(DMI_PRODUCT_NAME, "NC210/NC110"),
+			DMI_MATCH(DMI_BOARD_NAME, "NC210/NC110"),
+		},
+		.callback = dmi_check_cb,
+	},
 	{ },
 };
 MODULE_DEVICE_TABLE(dmi, samsung_dmi_table);
